gapic/src/main/com/google/gapid/perfetto/views/VulkanEventPanel.java

```diff
@@ -0,0 +1,203 @@
+  private static final double HOVER_MARGIN = 10;
+  private static final double HOVER_PADDING = 4;
+  private static final int BOUNDING_BOX_LINE_WIDTH = 3;
+
+  private final GpuInfo.VkEvent vkEvent;
+  protected final VulkanEventTrack track;
+
+  protected double mouseXpos, mouseYpos;
+  protected String hoveredName;
+  protected Size hoveredSize = Size.ZERO;
+
+  public VulkanEventPanel(State state, GpuInfo.VkEvent vkEvent, VulkanEventTrack track) {
+    super(state);
+    this.vkEvent = vkEvent;
+    this.track = track;
+  }
+
+  @Override
+  public VulkanEventPanel copy() {
+    return new VulkanEventPanel(state, vkEvent, track);
+  }
+
+  @Override
+  public String getTitle() {
+    return vkEvent.getDisplay();
+  }
+
+  @Override
+  public double getHeight() {
+    return SLICE_Y + vkEvent.maxDepth * SLICE_HEIGHT;
+  }
+
+  @Override
+  protected void renderTrack(RenderContext ctx, Repainter repainter, double w, double h) {
+    ctx.trace("VulkanEvents", () -> {
+      VulkanEventTrack.Data data = track.getData(state.toRequest(), onUiThread(repainter));
+      drawLoading(ctx, data, state, h);
+
+      if (data == null) {
+        return;
+      }
+
+      TimeSpan visible = state.getVisibleTime();
+      Selection<Long> selected = state.getSelection(Selection.Kind.VulkanEvent);
+      List<Integer> visibleSelected = Lists.newArrayList();
+
+      for (int i = 0; i < data.starts.length; i++) {
+        long tStart = data.starts[i];
+        long tEnd = data.ends[i];
+        int depth = data.depths[i];
+
+        if (tEnd <= visible.start || tStart >= visible.end) {
+          continue;
+        }
+        double rectStart = state.timeToPx(tStart);
+        double rectWidth = Math.max(1, state.timeToPx(tEnd) - rectStart);
+        double y = SLICE_Y + depth * SLICE_HEIGHT;
+
+        ctx.setBackgroundColor(VulkanEventTrack.getColor(data.names[i]));
+        ctx.fillRect(rectStart, y, rectWidth, SLICE_HEIGHT);
+
+        if (selected.contains(data.ids[i])) {
+          visibleSelected.add(i);
+        }
+
+        // Don't render text when we have less than 7px to play with.
+        if (rectWidth < 7) {
+          continue;
+        }
+
+        ctx.setForegroundColor(colors().textInvertedMain);
+        ctx.drawText(
+            Fonts.Style.Normal, data.names[i], rectStart + 2, y + 2, rectWidth - 4, SLICE_HEIGHT - 4);
+
+      }
+
+      // Draw bounding rectangles after all the slices are rendered, so that the border is on the top.
+      for (int index : visibleSelected) {
+        ctx.setForegroundColor(VulkanEventTrack.getBorderColor(data.names[index]));
+        double rectStart = state.timeToPx(data.starts[index]);
+        double rectWidth = Math.max(1, state.timeToPx(data.ends[index]) - rectStart);
+        double depth = data.depths[index];
+        ctx.drawRect(rectStart, SLICE_Y + depth * SLICE_HEIGHT, rectWidth, SLICE_HEIGHT, BOUNDING_BOX_LINE_WIDTH);
+
+        double mid = rectStart + rectWidth / 2;
```

Comment:
Just my opinion, so you might want to ask others about this.  But it would be more accurate to start the arrow at the beginning of the slice.  Otherwise, the render stages can start before the arrow.  Although I am not sure if that will look good.

Reply:
Thanks for suggestion! But personally I think it looks alright to grow arrow from the middle. Will keep it and mark as resolved for now, we can go back to change it in the future if more people raise the concern.